ghostscript: wrap binaries to add correct PATH

SharzyL opened the pull request:
###### Description of changes

For scripts in `${ghostscript}/bin`, they are not properly wrapped to put the binaries in PATH. When running them without packages in PATH, the scripts complain command not found. For example,

```
env -i nix shell p#ghostscript_headless -c ps2pdf
/nix/store/qg3n3ncn0hsi94ljr20m7f5w4kpxhqxg-ghostscript-9.56.1/bin/ps2pdf: line 8: dirname: command not found
/nix/store/qg3n3ncn0hsi94ljr20m7f5w4kpxhqxg-ghostscript-9.56.1/bin/ps2pdf14: line 3: dirname: command not found
/nix/store/qg3n3ncn0hsi94ljr20m7f5w4kpxhqxg-ghostscript-9.56.1/bin/ps2pdfwr: line 7: dirname: command not found
/nix/store/qg3n3ncn0hsi94ljr20m7f5w4kpxhqxg-ghostscript-9.56.1/bin/ps2pdfwr: line 24: basename: command not found
```

We wrapped these binaries to prefix the necessary packages to PATH.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

SharzyL committed:
ghostscript: wrap binaries to add correct PATH

SharzyL commented:
retarget staging now

drupol reviewed (commented):
This seems to be a legit change, but I would like to have the confirmation of more experienced reviewers.

drupol commented:
@xworld21 @apfelkuchen6 Can you help me reviewing this PR ?

xworld21 commented:
Uhm, doesn't texlive depend on ghostscript already?

I suppose this can work in a pinch, but it should definitely be revised to depend directly on `dvips`. Passing an empty combine is almost always wrong – either wasteful (the required package can be used directly) or broken (e.g. because of missing fonts).

apfelkuchen6 commented on pkgs/misc/ghostscript/default.nix:
```diff
@@ -26,6 +27,7 @@ let
       mv -v * "$out/"
     '';
   };
+  dvips = texlive.combine { inherit (texlive) scheme-infraonly dvips; };
```
I don't think this is helpful. `dvi` files don't contain any embedded fonts. dvips looks up font files (and other external files) using kpathsea and embeds them into the `ps` output. But if you take dvips from such a minimal texlive installation, this font lookup will always fail! So you won't be able to convert any nonempty file (ok, there might be some edge cases using only specials that work, but these won't appear in the real world ever).

Of course using `texlive.combined.scheme-full` here is ridiculous, so I think it's better to not include dvips in the wrapper at all. Then the user can at least provide their own dvips program (with the neccessary fonts etc).

SharzyL committed:
ghostscript: remove dvipdf wrapper

@apfelkuchen6 points out that “dvips looks up font files (and other
external files) using kpathsea and embeds them into the ps output.
But if you take dvips from such a minimal texlive installation, this
font lookup will always fail”. Since it is also ridiculous to use
texlive.combined.scheme-full, we do not wrap dvips and leaves the users
to provide their own dvips.

apfelkuchen6 reviewed (approved):
The changes look reasonable to me now.

There are failing tests on darwin, these also failed before, right?

xworld21 commented on pkgs/misc/ghostscript/default.nix:
```diff
@@ -139,7 +145,23 @@ stdenv.mkDerivation rec {
+    done
+
+    for bin in $out/bin/ps2* ; do
+      wrapProgram $bin --prefix PATH : ${lib.makeBinPath [ coreutils ]}:$out/bin
```
Have you checked that `$0` is interpreted correctly? I believe `wrapProgram` breaks `$0` for non-binary executables. I suspect in this case it would only break the help output, showing e.g. `.ps2pdf-wrapped` instead of `ps2pdf`. But `$0` is also used to find the sibling executables.

(By the way, for many of these scripts, a single `substituteInPlace` of the entire block `GS_EXECUTABLE= ... GS_EXECUTABLE="$gs"` with `GS_EXECUTABLE='${out}/bin/gs'` is enough, and possibly more appropriate than a wrapper.)

SharzyL replied:
Yes, help output is showing `.ps2pdf-wrapped` since it is using `basename $0`.

Substituting `GS_EXECUTABLE='${out}/bin/gs'` seems not enough because we also need to bind some PATH from coreutils, gawk.

xworld21 replied:
> Substituting `GS_EXECUTABLE='${out}/bin/gs'` seems not enough because we also need to bind some PATH from coreutils, gawk.

I meant, for quite a few scripts, the only coreutils usage is around `GS_EXECUTABLE`, so if you patch that away, there is no need for a wrapper. Other scripts do use gawk, basename, etc. so I agree with wrapping for those of course.
